tooling/nargo_cli/src/cli/test_cmd.rs

```diff
@@ -68,258 +94,410 @@ pub(crate) fn run(args: TestCommand, config: NargoConfig) -> Result<(), CliError
         None => FunctionNameMatch::Anything,
     };
 
-    // Configure a thread pool with a larger stack size to prevent overflowing stack in large programs.
-    // Default is 2MB.
-    let pool = rayon::ThreadPoolBuilder::new().stack_size(4 * 1024 * 1024).build().unwrap();
-    let test_reports: Vec<Vec<(String, TestStatus)>> = pool.install(|| {
-        workspace
-            .into_iter()
-            .par_bridge()
-            .map(|package| {
-                run_tests::<Bn254BlackBoxSolver>(
-                    &workspace_file_manager,
-                    &parsed_files,
-                    package,
-                    pattern,
-                    args.show_output,
-                    args.oracle_resolver.as_deref(),
-                    Some(workspace.root_dir.clone()),
-                    Some(package.name.to_string()),
-                    &args.compile_options,
-                )
-            })
-            .collect::<Result<_, _>>()
-    })?;
-    let test_report: Vec<(String, TestStatus)> = test_reports.into_iter().flatten().collect();
-
-    if test_report.is_empty() {
-        match &pattern {
-            FunctionNameMatch::Exact(pattern) => {
-                return Err(CliError::Generic(
-                    format!("Found 0 tests matching input '{pattern}'.",),
-                ))
+    let num_threads = args.test_threads.unwrap_or_else(rayon::current_num_threads);
+
+    let runner = TestRunner {
+        file_manager: &file_manager,
+        parsed_files: &parsed_files,
+        workspace,
+        args: &args,
+        pattern,
+        num_threads,
+    };
+    runner.run()
+}
+
+struct TestRunner<'a> {
+    file_manager: &'a FileManager,
+    parsed_files: &'a ParsedFiles,
+    workspace: Workspace,
+    args: &'a TestCommand,
+    pattern: FunctionNameMatch<'a>,
+    num_threads: usize,
+}
+
+impl<'a> TestRunner<'a> {
+    fn run(&self) -> Result<(), CliError> {
+        // First compile all packages and collect their tests
+        let package_test_results = self.collect_packages_tests();
+
+        // Now gather all tests and how many are per packages
+        let mut tests = Vec::new();
+        let mut test_count_per_package = HashMap::new();
+
+        for (package_name, package_tests) in package_test_results {
+            let package_tests = package_tests?;
+            test_count_per_package.insert(package_name, package_tests.len());
+            tests.extend(package_tests);
```

Comment:
I would use a `BTreeMap` instead of `HashMap` in all these cases, to fix the iteration order. One is to provide a more deterministic output among packages, and the other to make absolutely sure that the order we see while iterating `package_test_results`, which is going to be the order in which tests go into `tests`, is exactly going to be reflected in `test_count_per_package` as well, which we will iterate over when waiting for results to appear in that order. Or use a `Vec` for `test_count_per_package`.

Reply:
Great idea!